src/main/java/homer/controller/ControllerImpl.java

```diff
@@ -0,0 +1,82 @@
+import homer.api.DeviceView;
+import homer.api.ToggleableDevice;
+import homer.controller.command.Command;
+import homer.controller.command.ConnectDevice;
+import homer.controller.command.DisconnectAllDevices;
+import homer.controller.command.DisconnectDevice;
+import homer.controller.command.ToggleDevice;
+import homer.controller.command.UpdateDeviceState;
+import homer.model.outlets.OutletFactory;
+
+/**
+ * Controller Implementation.
+ */
+public final class ControllerImpl implements Controller {
+
+    private final Map<DeviceId, Device<?>> devices = new LinkedHashMap<>();
+    private final List<Command> commands = new LinkedList<>();
+
+    @Override
+    public Set<DeviceView<?>> getDevices() {
+        throw new UnsupportedOperationException("Unimplemented method 'getDevices'");
+    }
+
+    @Override
+    public void receiveCommand(final Command command) {
+        commands.add(command);
+    }
+
+    /**
+     * this should be called by the scheduler, which is not implemented yet.
+     * @param command
+     */
+    @SuppressWarnings("PMD.UnusedPrivateMethod") 
+    private void execute(final Command command) {
+        if (command instanceof ConnectDevice) { // TODO still not finished
+            final ConnectDevice connect = (ConnectDevice) command;
+            final String deviceType = connect.deviceType();
+            final DeviceId deviceId = new DeviceIdImpl();
+            devices.put(deviceId, OutletFactory.cOutlet(new DeviceInfoImpl(deviceId, deviceType), 0));
+        } else if (command instanceof DisconnectDevice) {
+            final DisconnectDevice remove = (DisconnectDevice) command;
+            devices.remove(remove.deviceId());
+        } else if (command instanceof DisconnectAllDevices) {
+            devices.clear();
+        } else if (command instanceof UpdateDeviceState) {
+            final UpdateDeviceState updateState = (UpdateDeviceState) command;
+            final Device<?> targetDevice = devices.get(updateState.deviceId());
+            if (targetDevice instanceof AdjustableDevice) {
+                final Class<?> adjustableClass = targetDevice.getClass();
+                final ParameterizedType parameterizedType = (ParameterizedType) adjustableClass.getGenericSuperclass();
+                final Type type = parameterizedType.getActualTypeArguments()[0];
+                if (type.equals(updateState.state().getClass()))  {
+                    final AdjustableDevice adjustableDevice = (AdjustableDevice) targetDevice;
+                    adjustableDevice.setState(updateState.state());
+                }
+            } 
+        } else if (command instanceof ToggleDevice) {
+            final Device<?> targetDevice = devices.get(((ToggleDevice) command).deviceId());
+            if (targetDevice instanceof ToggleableDevice) {
+                ((ToggleableDevice<?>) targetDevice).toggle();
+            }
+        } else {
+            throw new IllegalArgumentException();
+        }
+    }
```

Comment:
Unsure of how the command pattern is implemented here.
I will make my own for the scheduler.

Reply:
yeah, that's perfectly fine. this might not even be the final version, but i think for now it's good